src/use-swr.ts

```diff
@@ -213,149 +232,93 @@ function useSWR<Data = any, Error = any>(
+
+  // A revalidation must be triggered when mounted if:
+  // - `revalidateOnMount` is explicitly set to `true`.
+  // - Suspense mode and there's stale data for the inital render.
+  // - Not suspense mode and there is no `initialData`.
+  const shouldRevalidateOnMount = () => {
+    if (config.revalidateOnMount !== undefined) return config.revalidateOnMount
+
+    return config.suspense
+      ? !initialMountedRef.current && data !== undefined
+      : config.initialData === undefined
   }
 
-  const initialData = resolveData()
-  const initialError = cache.get(keyErr)
-  const initialIsValidating = resolveIsValidating()
-
-  // if a state is accessed (data, error or isValidating),
-  // we add the state to dependencies so if the state is
-  // updated in the future, we can trigger a rerender
-  const stateDependencies = useRef({
-    data: false,
-    error: false,
-    isValidating: false
-  })
-  const stateRef = useRef({
-    data: initialData,
-    error: initialError,
-    isValidating: initialIsValidating
-  })
-
-  // display the data label in the React DevTools next to SWR hooks
-  useDebugValue(stateRef.current.data)
-
-  const rerender = useState<unknown>({})[1]
+  // Resolve the current validating state.
+  const resolveValidating = () => {
+    if (!key) return false
+    if (cache.get(keyValidating)) return true
 
-  let dispatch = useCallback(
-    (payload: Action<Data, Error>) => {
-      let shouldUpdateState = false
-      for (let k in payload) {
-        // @ts-ignore
-        if (stateRef.current[k] === payload[k]) {
-          continue
-        }
-        // @ts-ignore
-        stateRef.current[k] = payload[k]
-        // @ts-ignore
-        if (stateDependencies.current[k]) {
-          shouldUpdateState = true
-        }
-      }
-
-      if (shouldUpdateState) {
-        // if component is unmounted, should skip rerender
-        // if component is not mounted, should skip rerender
-        if (unmountedRef.current || !initialMountedRef.current) return
-        rerender({})
-      }
-    },
-    // config.suspense isn't allowed to change during the lifecycle
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-    []
-  )
-
-  // error ref inside revalidate (is last request errored?)
-  const unmountedRef = useRef(false)
-  const keyRef = useRef(key)
-
-  // check if component is mounted in suspense mode
-  const initialMountedRef = useRef(false)
+    // If it's not mounted yet and it should revalidate on mount, revalidate.
+    return !initialMountedRef.current && shouldRevalidateOnMount()
+  }
+  const isValidating = resolveValidating()
 
   // do unmount check for callbacks
-  const eventsCallback = useCallback(
-    (event, ...params) => {
+  // if key changed during the revalidation, old dispatch and config callback should not take effect.
+  const safeCallback = useCallback(
+    (callback: () => void) => {
       if (unmountedRef.current) return
-      if (!initialMountedRef.current) return
```

Comment:
I think we still need to this initialMountedRef check for config callback

https://codesandbox.io/s/swr-787-165v7

Reply:
Good catch, thanks!